### PR TITLE
CEDS-1480 Move the DMS mapping from FE Exports

### PR DESCRIPTION
**Pull request by PauloPortugal**
* Use “plural” to denote the declarations resources/endpoints
* Add a country verification service

### CODE REVIEW

File: app/uk/gov/hmrc/exports/models/Choice.scala [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.models
+
+case class Choice(value: String)
+
+object Choice {
+  object AllowedChoiceValues {
+    val SupplementaryDec = "SMP"
```
Comment (bambuchaAdm) [inline app/uk/gov/hmrc/exports/models/Choice.scala]: Are we able to move this as case'objects extending same trait/abstract class ?

File: app/uk/gov/hmrc/exports/models/declaration/ExportItem.scala [hunk 2]
```diff
@@ -39,23 +42,26 @@ object AdditionalFiscalReferences {
...
+  descriptionOfGoods: String,
+  cusCode: Option[String],
+  unDangerousGoodsCode: Option[String],
+  statisticalValue: String
 )
 object ItemType {
   implicit val format: OFormat[ItemType] = Json.format[ItemType]
+
+  object IdentificationTypeCodes {
+    val CombinedNomenclatureCode = "TSP"
```
Comment (bambuchaAdm) [inline app/uk/gov/hmrc/exports/models/declaration/ExportItem.scala]: Same here - this looks like Enum - maybe seald trait/abstract class with some case objects in companion would be enough

File: conf/app.routes [hunk 3]
```diff
@@ -1,9 +1,9 @@
 # microservice specific routes
 
-POST        /v2/declaration                                  uk.gov.hmrc.exports.controllers.DeclarationController.create()
-GET         /v2/declaration                                  uk.gov.hmrc.exports.controllers.DeclarationController.findAll()
-GET         /v2/declaration/:id                              uk.gov.hmrc.exports.controllers.DeclarationController.findByID(id: String)
-DELETE      /v2/declaration/:id                              uk.gov.hmrc.exports.controllers.DeclarationController.deleteByID(id: String)
+POST        /v2/declarations                                 uk.gov.hmrc.exports.controllers.DeclarationController.create()
```
Comment (bambuchaAdm) [inline conf/app.routes]: :+1:

File: test/uk/gov/hmrc/exports/services/CountriesServiceSpec.scala [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.services
+
+import org.scalatestplus.play.PlaySpec
+import uk.gov.hmrc.exports.models.Country
+
+class CountriesServiceSpec extends PlaySpec {
```
Comment (bambuchaAdm) [inline test/uk/gov/hmrc/exports/services/CountriesServiceSpec.scala]: Why we have to get whole application for this spec ?

File: app/uk/gov/hmrc/exports/services/CountriesService.scala [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        case _ =>
+          throw new IllegalArgumentException("Could not read JSON array of countries from : " + jsonFile)
+      }
+
+    fromJsonFile.sortBy(_.countryName)
+  }
+
+  private def countryCode: String => String = cc => cc.split(":")(1).trim
+
+  val allCountries: List[Country] = countries
```
Comment (bambuchaAdm) [inline app/uk/gov/hmrc/exports/services/CountriesService.scala]: i am not sure if this assignment val to val is good. 

Maybe we should provide just functions to find country by code ?

File: app/uk/gov/hmrc/exports/models/Choice.scala (hunk 1 above)
Comment (ed0906) [reply app/uk/gov/hmrc/exports/models/Choice.scala]: Do we need this? It's not required on `ExportsDeclaration`

File: app/uk/gov/hmrc/exports/models/Country.scala [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.models
+
+import play.api.libs.json.Json
+
+case object Country {
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/models/Country.scala]: I don't think we need the `object` here as the formatter is never used

File: app/uk/gov/hmrc/exports/models/declaration/Locations.scala [hunk 7]
```diff
@@ -54,6 +55,15 @@ object WarehouseIdentification {
 case class OfficeOfExit(officeId: String, presentationOfficeId: Option[String], circumstancesCode: Option[String])
 object OfficeOfExit {
   implicit val format: OFormat[OfficeOfExit] = Json.format[OfficeOfExit]
+  def from(officeOfExitSupplementary: OfficeOfExitSupplementary): OfficeOfExit =
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/models/declaration/Locations.scala]: Why are these required? If the front end sends us an `OfficeOfExit` and the back end just needs an `OfficeOfExit` surely we can get rid of this?

File: app/uk/gov/hmrc/exports/models/declaration/officeOfExit/OfficeOfExitStandard.scala [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.models.declaration.officeOfExit
+
+import play.api.libs.json.Json
+import uk.gov.hmrc.exports.models.declaration.OfficeOfExit
+
+case class OfficeOfExitStandard(officeId: String, presentationOfficeId: Option[String], circumstancesCode: String)
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/models/declaration/officeOfExit/OfficeOfExitStandard.scala]: I'm interested to know why we need these

File: app/uk/gov/hmrc/exports/models/declaration/officeOfExit/OfficeOfExitStandard.scala [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+object OfficeOfExitStandard {
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/models/declaration/officeOfExit/OfficeOfExitStandard.scala]: I'm interested to know why we need these

File: app/uk/gov/hmrc/exports/models/declaration/officeOfExit/OfficeOfExitStandard.scala [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  implicit val format = Json.format[OfficeOfExitStandard]
+
+  def apply(officeOfExit: OfficeOfExit): OfficeOfExitStandard = OfficeOfExitStandard(
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/models/declaration/officeOfExit/OfficeOfExitStandard.scala]: I'm interested to know why we need these

File: app/uk/gov/hmrc/exports/models/declaration/officeOfExit/OfficeOfExitSupplementary.scala [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.models.declaration.officeOfExit
+
+import play.api.libs.json.Json
+import uk.gov.hmrc.exports.models.declaration.OfficeOfExit
+
+case class OfficeOfExitSupplementary(officeId: String)
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/models/declaration/officeOfExit/OfficeOfExitSupplementary.scala]: I'm interested to know why we need these

File: app/uk/gov/hmrc/exports/repositories/DeclarationRepository.scala [hunk 12]
```diff
@@ -33,6 +33,7 @@ class DeclarationRepository @Inject()(mc: ReactiveMongoComponent, appConfig: App
       ExportsDeclaration.Mongo.format,
       objectIdFormats
     ) {
+  def update(submittedDeclaration: Any) = ???
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/repositories/DeclarationRepository.scala]: I've implemented this now so you can revert this class

File: app/uk/gov/hmrc/exports/services/CountriesService.scala [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.services
+
+import play.api.libs.json._
+import uk.gov.hmrc.exports.models.Country
+
+object CountriesService {
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/services/CountriesService.scala]: If it's a service can it please be a class? Then we can use DI to mock it out in tests

File: app/uk/gov/hmrc/exports/services/DeclarationService.scala [hunk 14]
```diff
@@ -19,16 +19,22 @@ package uk.gov.hmrc.exports.services
...
+  wcoSubmissionService: WcoSubmissionService
+) {
+
+  def save(
+    declaration: ExportsDeclaration
+  )(implicit hc: HeaderCarrier, ec: ExecutionContext): Future[ExportsDeclaration] =
+    for {
+      saved <- declarationRepository.create(declaration)
+      _ <- wcoSubmissionService.submit(declaration)
+    } yield saved
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/services/DeclarationService.scala]: Nice and simple 👍

File: app/uk/gov/hmrc/exports/services/WcoMapperService.scala [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+import uk.gov.hmrc.exports.services.mapping.SubmissionMetaDataBuilder
+import wco.datamodel.wco.dec_dms._2.Declaration
+import wco.datamodel.wco.documentmetadata_dms._2.MetaData
+
+class WcoMapperService @Inject()(submissionMetadataBuilder: SubmissionMetaDataBuilder) {
+
+  def produceMetaData(exportsCacheModel: ExportsDeclaration): MetaData =
+    submissionMetadataBuilder.build(exportsCacheModel)
+
+  def declarationUcr(metaData: MetaData): Option[String] = {
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/services/WcoMapperService.scala]: Are these methods used?

File: app/uk/gov/hmrc/exports/services/WcoMapperService.scala [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+        .getGoodsShipment
+        .getUCR
+    )
+
+    ucr
+      .map(_.getTraderAssignedReferenceID.getValue)
+      .orElse(Some(""))
+  }
+
+  def declarationLrn(metaData: MetaData): Option[String] =
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/services/WcoMapperService.scala]: Are these methods used?

File: app/uk/gov/hmrc/exports/services/WcoMapperService.scala [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+    Option(
+      metaData.getAny
+        .asInstanceOf[JAXBElement[Declaration]]
+        .getValue
+        .getFunctionalReferenceID
+        .getValue
+    ).orElse(Some(""))
+
+  def toXml(metaData: MetaData) = {
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/services/WcoMapperService.scala]: Are these methods used?

File: app/uk/gov/hmrc/exports/services/mapping/CachingMappingHelper.scala [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+import uk.gov.hmrc.exports.models.declaration.ItemType.IdentificationTypeCodes.{
+  CUSCode,
+  CombinedNomenclatureCode,
+  NationalAdditionalCode,
+  TARICAdditionalCode
+}
+import uk.gov.hmrc.exports.models.declaration.{CommodityMeasure, ItemType}
+import uk.gov.hmrc.wco.dec._
+
+object CachingMappingHelper {
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/services/mapping/CachingMappingHelper.scala]: Is this used?

Comment (ed0906) [inline app/uk/gov/hmrc/exports/services/mapping/CachingMappingHelper.scala]: It'd be nice if this was a class rather than an object so the tests of things that use it don't need to care about what it does

File: app/uk/gov/hmrc/exports/services/mapping/declaration/ExitOfficeBuilder.scala [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+import uk.gov.hmrc.exports.models.Choice.AllowedChoiceValues
+import uk.gov.hmrc.exports.models.declaration.ExportsDeclaration
+import uk.gov.hmrc.exports.models.declaration.officeOfExit.{OfficeOfExitStandard, OfficeOfExitSupplementary}
+import uk.gov.hmrc.exports.services.mapping.ModifyingBuilder
+import wco.datamodel.wco.dec_dms._2.Declaration
+import wco.datamodel.wco.dec_dms._2.Declaration.ExitOffice
+import wco.datamodel.wco.declaration_ds.dms._2._
+
+class ExitOfficeBuilder @Inject()() extends ModifyingBuilder[ExportsDeclaration, Declaration] {
+  override def buildThenAdd(model: ExportsDeclaration, declaration: Declaration): Unit = model.choice match {
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/services/mapping/declaration/ExitOfficeBuilder.scala]: I think we can change this so it doesn't need to care about the different OfficeOfExit types.

Then we can get rid of the `Standard` and `Supplementary` classes

File: app/uk/gov/hmrc/exports/services/mapping/declaration/ExitOfficeBuilder.scala [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+        .foreach(declaration.setExitOffice)
+
+    case AllowedChoiceValues.SupplementaryDec =>
+      model.locations.officeOfExit
+        .map(OfficeOfExitSupplementary(_))
+        .map(createExitOfficeFromSupplementaryJourney)
+        .foreach(declaration.setExitOffice)
+  }
+
+  private def createExitOfficeFromStandardJourney(data: OfficeOfExitStandard): Declaration.ExitOffice = {
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/services/mapping/declaration/ExitOfficeBuilder.scala]: Looks to me like these two `create` methods are identical, we can just pass in an `OfficeOfExit` and remove the other

File: app/uk/gov/hmrc/exports/services/mapping/declaration/FunctionalReferenceIdBuilder.scala [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+      if (references.lrn.nonEmpty) {
+        declaration.setFunctionalReferenceID(build(references.lrn))
+      }
+    })
+  }
+}
+
+object FunctionalReferenceIdBuilder {
+
+  def build(functionalReferenceId: String): DeclarationFunctionalReferenceIDType = {
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/services/mapping/declaration/FunctionalReferenceIdBuilder.scala]: No need for this to be in an object, it can be moved into the class

File: app/uk/gov/hmrc/exports/services/mapping/declaration/TypeCodeBuilder.scala [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+    val typeCodeType = new DeclarationTypeCodeType()
+    dispatchLocation.foreach { data =>
+      typeCodeType.setValue(data.dispatchLocation + decType.additionalDeclarationType)
+    }
+    typeCodeType
+  }
+}
+object TypeCodeBuilder {
+
+  def build(codeType: String): DeclarationTypeCodeType = {
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/services/mapping/declaration/TypeCodeBuilder.scala]: This isnt used

File: app/uk/gov/hmrc/exports/services/mapping/goodsshipment/ExportCountryBuilder.scala [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+      goodsShipment.setExportCountry(createExportCountry(model))
+    }
+
+  private def isDefined(country: DestinationCountries): Boolean = country.countryOfDispatch.nonEmpty
+
+  private def createExportCountry(data: DestinationCountries): GoodsShipment.ExportCountry = {
+
+    val id = new ExportCountryCountryCodeType()
+    id.setValue(
+      CountriesService.allCountries
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/services/mapping/goodsshipment/ExportCountryBuilder.scala]: It'd be nice if we could  mock this countries call out

File: conf/app.routes (hunk 3 above)
Comment (ed0906) [reply conf/app.routes]: This will need updating in the frontend application.conf

File: test/uk/gov/hmrc/exports/services/CountriesServiceSpec.scala (hunk 4 above)
Comment (ed0906) [reply test/uk/gov/hmrc/exports/services/CountriesServiceSpec.scala]: It doesnt @bambuchaAdm

**Comment by ed0906** [comment]
I've left a few comments, it would be good to run through the journey in the front end to verify it works as planned

Also I think there's an end to end test for the mapping that may be worth bringing across

File: app/uk/gov/hmrc/exports/services/WcoSubmissionService.scala [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+
+    val metaData = wcoMapperService.produceMetaData(declaration)
+    val lrn = wcoMapperService.declarationLrn(metaData)
+    val ducr = wcoMapperService.declarationUcr(metaData)
+
+    customsDeclarationsConnector.submitDeclaration(declaration.eori, wcoMapperService.toXml(metaData)).flatMap {
+      case CustomsDeclarationsResponse(ACCEPTED, Some(conversationId)) =>
+        val newSubmission =
+          Submission(
+            eori = declaration.eori,
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/services/WcoSubmissionService.scala]: We could set the id on the submission to be the same as the declaration. Might make it easier for us to tie the two together. At the moment the default is used which is a new uuid

File: app/uk/gov/hmrc/exports/services/mapping/declaration/ExitOfficeBuilder.scala (hunk 20 above)
Comment (ed0906) [reply app/uk/gov/hmrc/exports/services/mapping/declaration/ExitOfficeBuilder.scala]: Fixed this for you @PauloPortugal

File: app/uk/gov/hmrc/exports/services/mapping/declaration/ExitOfficeBuilder.scala (hunk 19 above)
Comment (ed0906) [reply app/uk/gov/hmrc/exports/services/mapping/declaration/ExitOfficeBuilder.scala]: Fixed this for you @PauloPortugal

File: app/uk/gov/hmrc/exports/models/declaration/officeOfExit/OfficeOfExitSupplementary.scala (hunk 11 above)
Comment (ed0906) [reply app/uk/gov/hmrc/exports/models/declaration/officeOfExit/OfficeOfExitSupplementary.scala]: Fixed this for you @PauloPortugal

File: app/uk/gov/hmrc/exports/models/declaration/officeOfExit/OfficeOfExitStandard.scala (hunk 10 above)
Comment (ed0906) [reply app/uk/gov/hmrc/exports/models/declaration/officeOfExit/OfficeOfExitStandard.scala]: Fixed this for you @PauloPortugal

File: app/uk/gov/hmrc/exports/models/declaration/officeOfExit/OfficeOfExitStandard.scala (hunk 9 above)
Comment (ed0906) [reply app/uk/gov/hmrc/exports/models/declaration/officeOfExit/OfficeOfExitStandard.scala]: Fixed this for you @PauloPortugal

File: app/uk/gov/hmrc/exports/models/declaration/officeOfExit/OfficeOfExitStandard.scala (hunk 8 above)
Comment (ed0906) [reply app/uk/gov/hmrc/exports/models/declaration/officeOfExit/OfficeOfExitStandard.scala]: Fixed this for you @PauloPortugal

File: app/uk/gov/hmrc/exports/models/declaration/Locations.scala (hunk 7 above)
Comment (ed0906) [reply app/uk/gov/hmrc/exports/models/declaration/Locations.scala]: Fixed this for you @PauloPortugal

File: app/uk/gov/hmrc/exports/repositories/DeclarationRepository.scala (hunk 12 above)
Comment (ed0906) [reply app/uk/gov/hmrc/exports/repositories/DeclarationRepository.scala]: I've merged in master for you @PauloPortugal

File: app/uk/gov/hmrc/exports/models/Choice.scala (hunk 1 above)
Comment (ed0906) [reply app/uk/gov/hmrc/exports/models/Choice.scala]: Fixed

**Comment by ed0906** [comment]
I've addressed a few of my own comments for you (see commits above) and merged in Master. I'll let you decide what you think to the others 👍

File: test/uk/gov/hmrc/exports/services/CountriesServiceSpec.scala (hunk 4 above)
Comment (PauloPortugal) [reply test/uk/gov/hmrc/exports/services/CountriesServiceSpec.scala]: This has been addressed.

**Comment by ed0906** [comment]
PR for the front end URL change https://github.com/hmrc/customs-declare-exports-frontend/pull/414

File: app/uk/gov/hmrc/exports/models/Choice.scala (hunk 1 above)
Comment (PauloPortugal) [reply app/uk/gov/hmrc/exports/models/Choice.scala]: This is used on the Mappers.

File: test/uk/gov/hmrc/exports/services/CountriesServiceSpec.scala (hunk 4 above)
Comment (PauloPortugal) [reply test/uk/gov/hmrc/exports/services/CountriesServiceSpec.scala]: I have addressed this

File: app/uk/gov/hmrc/exports/models/declaration/ExportItem.scala (hunk 2 above)
Comment (PauloPortugal) [reply app/uk/gov/hmrc/exports/models/declaration/ExportItem.scala]: this has been addressed

File: app/uk/gov/hmrc/exports/models/Country.scala (hunk 6 above)
Comment (PauloPortugal) [reply app/uk/gov/hmrc/exports/models/Country.scala]: This has been addressed

File: app/uk/gov/hmrc/exports/services/CountriesService.scala (hunk 13 above)
Comment (PauloPortugal) [reply app/uk/gov/hmrc/exports/services/CountriesService.scala]: This has been addressed.

File: app/uk/gov/hmrc/exports/services/mapping/goodsshipment/ExportCountryBuilder.scala (hunk 23 above)
Comment (PauloPortugal) [reply app/uk/gov/hmrc/exports/services/mapping/goodsshipment/ExportCountryBuilder.scala]: this has been addressed

File: app/uk/gov/hmrc/exports/services/WcoSubmissionService.scala (hunk 24 above)
Comment (PauloPortugal) [reply app/uk/gov/hmrc/exports/services/WcoSubmissionService.scala]: Agreed. Let's do this as part of a ticket to unify the Submission into the Declaration

File: app/uk/gov/hmrc/exports/models/Choice.scala [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.exports.models
+import play.api.libs.json._
+
+sealed abstract class Choice(val value: String)
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/models/Choice.scala]: This would be a lot less code if we used Scala Enumerations e.g. 

https://github.com/hmrc/customs-declare-exports/blob/master/app/uk/gov/hmrc/exports/models/declaration/DeclarationStatus.scala

You can give them String Values  too e.g.

```
object Choice extends Enumeration {
  type Choice = Value
  implicit val format: Format[Choice.Value] = EnumJson.format(Choice)
  val Standard = Value("STD")
  val Supplementary = Value("SMP")
}
```

File: app/uk/gov/hmrc/exports/models/declaration/ExportItem.scala [hunk 26]
```diff
@@ -39,23 +42,41 @@ object AdditionalFiscalReferences {
...
 object ItemType {
   implicit val format: OFormat[ItemType] = Json.format[ItemType]
 }
 
-case class PackageInformation(
-  typesOfPackages: Option[String],
-  numberOfPackages: Option[Int],
-  shippingMarks: Option[String]
-)
+sealed abstract class IdentificationTypeCodes(val value: String)
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/models/declaration/ExportItem.scala]: Similar here, if we used Scala Enumerations this would be a lot less code

File: app/uk/gov/hmrc/exports/models/declaration/ExportsDeclaration.scala [hunk 27]
```diff
@@ -44,6 +45,51 @@ case class ExportsDeclaration(
 )
 
 object ExportsDeclaration {
+  
+  @deprecated("Please use apply with `Choice` class", since = "2019-08-14")
```
Comment (ed0906) [inline app/uk/gov/hmrc/exports/models/declaration/ExportsDeclaration.scala]: Cant we just change the existing usages rather than add a deprecated constructor?

File: app/uk/gov/hmrc/exports/repositories/DeclarationRepository.scala (hunk 12 above)
Comment (ed0906) [reply app/uk/gov/hmrc/exports/repositories/DeclarationRepository.scala]: Looks like it's been overwritten...

**Comment by ed0906** [comment]
I think we should have another discussion of the benefits of Merge Vs Rebase & Force Push. 
If we used Merge then this PR would have been a lot less painful (in my opinion)

File: app/uk/gov/hmrc/exports/repositories/DeclarationRepository.scala (hunk 12 above)
Comment (PauloPortugal) [reply app/uk/gov/hmrc/exports/repositories/DeclarationRepository.scala]: `rebase` vs `merge` sorry my fault there.

File: app/uk/gov/hmrc/exports/models/Choice.scala (hunk 25 above)
Comment (PauloPortugal) [reply app/uk/gov/hmrc/exports/models/Choice.scala]: As discussed we are going with a Sealed Trait. The added effort of writes and reads seem unnecessary when compared to an Enum, but the wider community tend to prefer sealed traits. Let's see where this takes us.

File: app/uk/gov/hmrc/exports/models/declaration/ExportItem.scala (hunk 26 above)
Comment (PauloPortugal) [reply app/uk/gov/hmrc/exports/models/declaration/ExportItem.scala]: as mentioned in another comment...

File: app/uk/gov/hmrc/exports/services/mapping/CachingMappingHelper.scala (hunk 18 above)
Comment (PauloPortugal) [reply app/uk/gov/hmrc/exports/services/mapping/CachingMappingHelper.scala]: Thsi has been addressed

**Comment by PauloPortugal** [comment]
> I think we should have another discussion of the benefits of Merge Vs Rebase & Force Push.
> If we used Merge then this PR would have been a lot less painful (in my opinion)

I agree with the discussion.
I should have not used rebase the minute you jumped on the branch, my fault there, because `merge` preserves history whereas `rebase` rewrites it.

For private branching, i.e, only one developer working on it rebasing makes a lot of sense. You get cleaner commits, structural logical commits and we should explain why we did something.
I am quite critical of the git commit hygiene in the project, as in my opinion, is substandard, as in, commits are treated as second class citizens when they should not be.

https://www.atlassian.com/git/tutorials/merging-vs-rebasing is great to understand the difference.

The golden rule is when on public branches, then rebase should not be the weapon of choice.
It worked well when I worked with @PatrykRudnicki on the same branch, but again he also `rebases` and understands the git command line.
